File: semana_7/ivan_peralta/MergeSortIterative.java
public class MergeSortIterative {

    public static void main(String[] args) {
        int[] arr = { 12, 11, 13, 5, 6, 7 };
        int n = arr.length;

        mergeSort(arr, n);
        printArray(arr);
    }

    static void mergeSort(int[] arr, int n) {
        int curr_size;
        int left_start;

        for (curr_size = 1; curr_size <= n - 1; curr_size = 2 * curr_size) {
            for (left_start = 0; left_start < n - 1; left_start += 2 * curr_size) {
                int mid = Math.min(left_start + curr_size - 1, n - 1);
                int right_end = Math.min(left_start + 2 * curr_size - 1, n - 1);

                merge(arr, left_start, mid, right_end);
            }
        }
    }

    static void merge(int[] arr, int left, int mid, int right) {
        int n1 = mid - left + 1;
        int n2 = right - mid;

        int[] L = new int[n1];
        int[] R = new int[n2];

        for (int i = 0; i < n1; ++i) {
            L[i] = arr[left + i];
        }
        for (int j = 0; j < n2; ++j) {
            R[j] = arr[mid + 1 + j];
        }

        int i = 0, j = 0;
        int k = left;
        while (i < n1 && j < n2) {
            if (L[i] <= R[j]) {
                arr[k] = L[i];
                i++;
            } else {
                arr[k] = R[j];
                j++;
            }
            k++;
        }

        while (i < n1) {
            arr[k] = L[i];
            i++;
            k++;
        }

        while (j < n2) {
            arr[k] = R[j];
            j++;
            k++;
        }
    }

    static void printArray(int[] arr) {
        for (int i : arr) {
            System.out.print(i + " ");
        }
        System.out.println();
    }
}

File: semana_7/BrunOrtega/mergesort_iterative.py
def mergesort_iterative(arr):
    n = len(arr)
    width = 1
    while width < n:
        left = 0
        while left < n:
            mid = left + width
            right = min(mid + width, n)
            arr[left:right] = merge(arr[left:mid], arr[mid:right])
            left = right
        width *= 2
    return arr

def merge(left, right):
    result = []
    while len(left) > 0 and len(right) > 0:
        if left[0] <= right[0]:
            result.append(left.pop(0))
        else:
            result.append(right.pop(0))
    result.extend(left)
    result.extend(right)
    return result

def main():
    arr = [5, 2, 8, 3, 1, 6, 4]
    arr = mergesort_iterative(arr)
    print("Sorted array:", arr)

if __name__ == "__main__":
    main()